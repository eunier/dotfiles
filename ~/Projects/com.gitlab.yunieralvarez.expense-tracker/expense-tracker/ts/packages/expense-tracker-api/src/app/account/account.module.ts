import { Module } from '@nestjs/common';
import { AccountRepositoryProvider } from '~/app/account/providers/account-repository/account-repository.provider';

@Module({
	providers: [AccountRepositoryProvider],
})
export class AccountModule {}
