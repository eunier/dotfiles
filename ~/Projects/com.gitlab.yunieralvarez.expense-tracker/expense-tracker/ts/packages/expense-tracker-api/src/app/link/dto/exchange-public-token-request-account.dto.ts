import { IsString } from 'class-validator';

export class ExchangePublicTokenRequestAccount {
	@IsString()
	public readonly plaidId: string;

	@IsString()
	public readonly name: string;

	@IsString()
	public readonly subtype: string;

	@IsString()
	public readonly mask: string;
}
