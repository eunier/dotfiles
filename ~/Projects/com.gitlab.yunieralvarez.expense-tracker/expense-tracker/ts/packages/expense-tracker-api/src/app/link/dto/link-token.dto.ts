import { PlaidLib } from "~/app/plaid/libs/plaid.lib";

export class LinkTokenDto {
	public readonly linkToken: string;

	constructor(linkToken: string) {
		this.linkToken = linkToken;
	}

	public static fromPlaidLinkTokenCreateResponse(response: PlaidLib.LinkTokenCreateResponse) {
		return new LinkTokenDto(response.link_token);
	}
}
