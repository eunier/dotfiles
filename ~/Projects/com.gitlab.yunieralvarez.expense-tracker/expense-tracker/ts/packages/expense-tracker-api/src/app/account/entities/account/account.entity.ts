import {
	Collection,
	Entity,
	EntityRepositoryType,
	ManyToOne,
	OneToMany,
	Property,
} from '@mikro-orm/core';
import { AccountRepository } from '~/app/account/repositories/account/account.repository';
import { ExchangePublicTokenRequestAccount } from '~/app/link/dto/exchange-public-token-request-account.dto';
import { BaseEntity } from '~/app/orm/entities/base/base.entity';
import { SourceEntity } from '~/app/source/entities/source/source.entity';
import { TransactionEntity } from '~/app/transaction/entities/transaction/transaction.entity';

@Entity({
	tableName: AccountEntity.metadata.tableName,
	repository: AccountEntity.metadata.repositoryFn,
})
export class AccountEntity extends BaseEntity {
	public static metadata = {
		repository: AccountRepository,
		repositoryFn: () => AccountEntity.metadata.repository,
		tableName: 'account',
	};

	[EntityRepositoryType]? = AccountEntity.metadata.repository;

	@Property()
	public plaidId: string;

	@Property()
	public name: string;

	@Property()
	public subtype: string;

	@Property()
	public mask: string;

	@ManyToOne(() => SourceEntity)
	public source: SourceEntity;

	@OneToMany(
		() => TransactionEntity,
		(transaction) => transaction.account,
	)
	public transactions = new Collection<TransactionEntity>(this);

	public constructor(
		args?: Partial<
			Pick<
				AccountEntity,
				'id' | 'mask' | 'name' | 'plaidId' | 'source' | 'subtype'
			>
		>,
	) {
		super();
		if (args) Object.assign(this, args);
	}

	public static fromExchangePublicTokenRequestAccount(
		request: ExchangePublicTokenRequestAccount,
	) {
		return new AccountEntity({
			mask: request.mask,
			name: request.name,
			plaidId: request.plaidId,
			subtype: request.subtype,
		});
	}
}
