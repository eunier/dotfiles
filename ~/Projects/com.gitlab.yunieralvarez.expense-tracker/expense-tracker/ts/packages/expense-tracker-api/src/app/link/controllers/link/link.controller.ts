import { Body, Controller, Get, Logger, Post } from '@nestjs/common';
import { ApiTags } from '@nestjs/swagger';
import { ActiveUser } from '~/app/auth/decorators/active-user.decorator';
import { ActiveUserData } from '~/app/auth/models/active-user-data.model';
import { ExchangePublicTokenRequestDto } from '~/app/link/dto/exchange-public-token-request.dto';
import { LinkService } from '~/app/link/services/link/link.service';

@ApiTags('Link')
@Controller('link')
export class LinkController {
	private readonly logger = new Logger(LinkController.name);

	public constructor(private readonly linkService: LinkService) {}

	@Get('token')
	public async getLinkToken(@ActiveUser('sub') userId: ActiveUserData['sub']) {
		this.logger.log('Creating Plaid Link token...');
		const response = await this.linkService.createLinkToken(userId);

		this.logger.debug(
			`Created Plaid Link token: ${JSON.stringify(response, null, 2)}`,
		);

		return response;
	}

	@Post('token/public/exchange')
	public async exchangePublicToken(
		@Body() exchangePublicTokenDto: ExchangePublicTokenRequestDto,
		@ActiveUser('sub') userId: ActiveUserData['sub'],
	) {
		this.logger.debug(
			`Exchanging '${exchangePublicTokenDto.publicToken}' public token for access token...`,
		);

		await this.linkService.exchangePublicToken(exchangePublicTokenDto, userId);

		this.logger.debug(
			`Exchanged '${exchangePublicTokenDto.publicToken}' public token for access token`,
		);
	}
}
