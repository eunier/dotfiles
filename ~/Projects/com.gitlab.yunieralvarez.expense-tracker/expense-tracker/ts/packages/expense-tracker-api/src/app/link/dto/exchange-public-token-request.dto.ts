import { ApiProperty } from '@nestjs/swagger';
import { IsArray, IsString } from 'class-validator';
import { ExchangePublicTokenRequestAccount } from '~/app/link/dto/exchange-public-token-request-account.dto';

export class ExchangePublicTokenRequestDto {
	@ApiProperty({
		description:
			'This is the public token returned by the Plaid Link on the UI.',
	})
	@IsString()
	public readonly publicToken: string;

	@ApiProperty({
		description: 'The institution for which the user selected the account.',
	})
	@IsString()
	public readonly institutionId: string;

	@IsArray()
	public readonly accounts: ExchangePublicTokenRequestAccount[];
}
