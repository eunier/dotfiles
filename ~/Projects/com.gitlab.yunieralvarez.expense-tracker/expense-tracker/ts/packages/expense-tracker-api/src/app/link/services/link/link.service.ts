import { EntityManager } from '@mikro-orm/postgresql';
import { Injectable, Logger, NotFoundException } from '@nestjs/common';
import { AccountEntity } from '~/app/account/entities/account/account.entity';
import { InstitutionService } from '~/app/institution/services/institution/institution.service';
import { ExchangePublicTokenRequestDto } from '~/app/link/dto/exchange-public-token-request.dto';
import { LinkTokenDto } from '~/app/link/dto/link-token.dto';
import { PlaidService } from '~/app/plaid/services/plaid/plaid.service';
import { SourceEntity } from '~/app/source/entities/source/source.entity';
import { UserRepository } from '~/app/user/repositories/user/user.repositories';

@Injectable()
export class LinkService {
	private readonly logger = new Logger(LinkService.name);

	public constructor(
		private readonly entityManager: EntityManager,
		private readonly institutionService: InstitutionService,
		private readonly plaidService: PlaidService,
		private readonly userRepository: UserRepository,
	) {}

	/**
	 * Create a Plaid link token.
	 * @param userId The user id.
	 * @returns Plaid link token.
	 */
	public async createLinkToken(userId: string) {
		const response = await this.plaidService.createLinkToken(userId);
		return LinkTokenDto.fromPlaidLinkTokenCreateResponse(response);
	}

	/**
	 * Exchange a Plaid public token for a access token.
	 * @param exchangePublicTokenDto The request dto.
	 * @param userId The user id.
	 */
	public async exchangePublicToken(
		dto: ExchangePublicTokenRequestDto,
		userId: string,
	) {
		const exchangePublicTokenResponse =
			await this.plaidService.exchangePublicToken(dto.publicToken);

		const source = SourceEntity.fromExchangePublicTokenResponse(
			exchangePublicTokenResponse,
		);

		const institution = await this.institutionService.getAndUpsertInstitution(
			dto.institutionId,
		);

		const user = await this.userRepository.findOne({ id: userId });

		if (!user) {
			this.logger.error(`User with id ${userId} not found`);

			throw new NotFoundException(`User with id ${userId} not found`, {
				description: 'The user ID provided does not exist in the database.',
			});
		}

		source.user = user;
		source.institution = institution;

		const accounts = dto.accounts.map(
			AccountEntity.fromExchangePublicTokenRequestAccount,
		);

		for (const account of accounts) {
			account.source = source;
		}

		this.logger.verbose(`Saving ${accounts.length} account(s)`);
		await this.entityManager.persistAndFlush(accounts);
	}
}
