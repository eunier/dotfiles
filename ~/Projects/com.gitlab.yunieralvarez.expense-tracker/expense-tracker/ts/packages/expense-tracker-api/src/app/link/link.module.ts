import { Module } from '@nestjs/common';
import { ConfigModule } from '~/app/config/config.module';
import { InstitutionEntity } from '~/app/institution/entities/institution/institution.entity';
import { InstitutionModule } from '~/app/institution/institution.module';
import { LinkController } from '~/app/link/controllers/link/link.controller';
import { LinkService } from '~/app/link/services/link/link.service';
import { OrmModule } from '~/app/orm/orm.module';
import { PlaidModule } from '~/app/plaid/plaid.module';
import { UserEntity } from '~/app/user/entities/user/user.entity';
import { UserModule } from '~/app/user/user.module';

@Module({
	imports: [
		OrmModule.forFeature([UserEntity, InstitutionEntity]),
		ConfigModule,
		InstitutionModule,
		PlaidModule,
		UserModule,
	],
	controllers: [LinkController],
	providers: [LinkService],
})
export class LinkModule {}
