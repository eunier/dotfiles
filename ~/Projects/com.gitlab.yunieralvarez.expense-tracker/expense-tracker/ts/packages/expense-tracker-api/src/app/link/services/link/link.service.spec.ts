import { faker } from '@faker-js/faker/.';
import { EntityManager } from '@mikro-orm/postgresql';
import { NotFoundException } from '@nestjs/common';
import { Test, TestingModule } from '@nestjs/testing';
import { AccountEntity } from '~/app/account/entities/account/account.entity';
import { InstitutionEntity } from '~/app/institution/entities/institution/institution.entity';
import { InstitutionService } from '~/app/institution/services/institution/institution.service';
import { ExchangePublicTokenRequestDto } from '~/app/link/dto/exchange-public-token-request.dto';
import { LinkTokenDto } from '~/app/link/dto/link-token.dto';
import { LinkService } from '~/app/link/services/link/link.service';
import { PlaidLib } from '~/app/plaid/libs/plaid.lib';
import { ExchangePublicTokenResponse } from '~/app/plaid/models/exchange-public-token-response/exchange-public-token-response.model';
import { PlaidService } from '~/app/plaid/services/plaid/plaid.service';
import { UserEntity } from '~/app/user/entities/user/user.entity';
import { UserRepository } from '~/app/user/repositories/user/user.repositories';
import { Utils } from '~/app/utils/services/utils/utils.service';

beforeEach(() => {
	jest.useFakeTimers();
	jest.setSystemTime(new Date(0).getTime());

	faker.seed(
		Utils.genSeedFromString(
			expect.getState().currentTestName ?? LinkService.name,
		),
	);
});

describe('LinkService', () => {
	let linkService: LinkService;
	let plaidService: PlaidService;
	let institutionService: InstitutionService;
	let entityManager: EntityManager;
	let userRepository: UserRepository;

	beforeEach(async () => {
		const module: TestingModule = await Test.createTestingModule({
			providers: [
				LinkService,
				{
					provide: PlaidService,
					useValue: {
						createLinkToken: jest.fn(),
						exchangePublicToken: jest.fn(),
					} satisfies Partial<PlaidService>,
				},
				{
					provide: UserRepository,
					useValue: { findOne: jest.fn() } satisfies Partial<UserRepository>,
				},
				{
					provide: EntityManager,
					useValue: {
						persistAndFlush: jest.fn(),
					} satisfies Partial<EntityManager>,
				},
				{
					provide: InstitutionService,
					useValue: {
						getAndUpsertInstitution: jest.fn(),
					} satisfies Partial<InstitutionService>,
				},
			],
		}).compile();

		linkService = module.get<LinkService>(LinkService);
		plaidService = module.get<PlaidService>(PlaidService);
		institutionService = module.get<InstitutionService>(InstitutionService);
		entityManager = module.get<EntityManager>(EntityManager);
		userRepository = module.get<UserRepository>(UserRepository);
	});

	it('should be defined', () => {
		expect(linkService).toBeDefined();
	});

	describe('createLinkToken', () => {
		it('should return a link token dto', async () => {
			jest.spyOn(plaidService, 'createLinkToken').mockResolvedValue({
				link_token: faker.string.nanoid(),
			} satisfies Partial<PlaidLib.LinkTokenCreateResponse> as PlaidLib.LinkTokenCreateResponse);

			const res = await linkService.createLinkToken(faker.string.nanoid());

			expect(res).toEqual({
				linkToken: 'z35b6q3t_qj0xpg-bt2uD',
			} satisfies Partial<LinkTokenDto>);
		});
	});

	describe('exchangePublicToken', () => {
		describe('when no user is found', () => {
			it('should throw a not found exception', async () => {
				jest.spyOn(plaidService, 'exchangePublicToken').mockResolvedValue({
					accessToken: faker.string.nanoid(),
					plaidId: faker.string.nanoid(),
					requestId: faker.string.nanoid(),
				} satisfies Partial<ExchangePublicTokenResponse> as ExchangePublicTokenResponse);

				jest
					.spyOn(institutionService, 'getAndUpsertInstitution')
					.mockResolvedValue({
						id: faker.string.nanoid(),
						name: faker.company.name(),
						plaidId: faker.string.nanoid(),
					} satisfies Partial<InstitutionEntity> as InstitutionEntity);

				const entityManagerPersistAndFlushSpy = jest.spyOn(
					entityManager,
					'persistAndFlush',
				);

				const userId = faker.string.nanoid();

				const dto: ExchangePublicTokenRequestDto = {
					publicToken: faker.string.nanoid(),
					institutionId: faker.string.nanoid(),
					accounts: [
						{
							mask: faker.string.numeric(4),
							name: faker.company.name(),
							plaidId: faker.string.nanoid(),
							subtype: faker.string.nanoid(),
						},
						{
							mask: faker.string.numeric(4),
							name: faker.company.name(),
							plaidId: faker.string.nanoid(),
							subtype: faker.string.nanoid(),
						},
					],
				};

				try {
					await linkService.exchangePublicToken(dto, userId);
					fail('Expected NotFoundException to be thrown');
				} catch (err) {
					expect(err).toBeInstanceOf(NotFoundException);

					expect((err as NotFoundException).message).toEqual(
						`User with id ${userId} not found`,
					);
				}
			});
			it('should not call persistAndFlush', async () => {
				jest.spyOn(plaidService, 'exchangePublicToken').mockResolvedValue({
					accessToken: faker.string.nanoid(),
					plaidId: faker.string.nanoid(),
					requestId: faker.string.nanoid(),
				} satisfies Partial<ExchangePublicTokenResponse> as ExchangePublicTokenResponse);

				jest
					.spyOn(institutionService, 'getAndUpsertInstitution')
					.mockResolvedValue({
						id: faker.string.nanoid(),
						name: faker.company.name(),
						plaidId: faker.string.nanoid(),
					} satisfies Partial<InstitutionEntity> as InstitutionEntity);

				const entityManagerPersistAndFlushSpy = jest.spyOn(
					entityManager,
					'persistAndFlush',
				);

				const userId = faker.string.nanoid();

				const dto: ExchangePublicTokenRequestDto = {
					publicToken: faker.string.nanoid(),
					institutionId: faker.string.nanoid(),
					accounts: [
						{
							mask: faker.string.numeric(4),
							name: faker.company.name(),
							plaidId: faker.string.nanoid(),
							subtype: faker.string.nanoid(),
						},
						{
							mask: faker.string.numeric(4),
							name: faker.company.name(),
							plaidId: faker.string.nanoid(),
							subtype: faker.string.nanoid(),
						},
					],
				};

				try {
					await linkService.exchangePublicToken(dto, userId);
					fail('Expected NotFoundException to be thrown');
				} catch (_) {}

				expect(entityManagerPersistAndFlushSpy).not.toHaveBeenCalled();
			});
		});

		describe('when user is found', () => {
			it('it should attached the user to the source', async () => {
				jest.spyOn(plaidService, 'exchangePublicToken').mockResolvedValue({
					accessToken: faker.string.nanoid(),
					plaidId: faker.string.nanoid(),
					requestId: faker.string.nanoid(),
				} satisfies Partial<ExchangePublicTokenResponse> as ExchangePublicTokenResponse);

				jest
					.spyOn(institutionService, 'getAndUpsertInstitution')
					.mockResolvedValue({
						id: faker.string.nanoid(),
						name: faker.company.name(),
						plaidId: faker.string.nanoid(),
					} satisfies Partial<InstitutionEntity> as InstitutionEntity);

				const userId = faker.string.nanoid();

				jest.spyOn(userRepository, 'findOne').mockResolvedValue({
					id: userId,
					email: faker.internet.email(),
					password: faker.string.nanoid(),
				} satisfies Partial<UserEntity> as UserEntity);

				const entityManagerPersistAndFlushSpy = jest.spyOn(
					entityManager,
					'persistAndFlush',
				);

				const dto: ExchangePublicTokenRequestDto = {
					publicToken: faker.string.nanoid(),
					institutionId: faker.string.nanoid(),
					accounts: [
						{
							mask: faker.string.numeric(4),
							name: faker.company.name(),
							plaidId: faker.string.nanoid(),
							subtype: faker.string.nanoid(),
						},
						{
							mask: faker.string.numeric(4),
							name: faker.company.name(),
							plaidId: faker.string.nanoid(),
							subtype: faker.string.nanoid(),
						},
					],
				};

				await linkService.exchangePublicToken(dto, userId);

				const entityManagerPersistAndFlushArgs = entityManagerPersistAndFlushSpy
					.mock.calls[0][0] as AccountEntity[];

				expect(entityManagerPersistAndFlushArgs).toMatchSnapshot();
			});
		});
	});
});
