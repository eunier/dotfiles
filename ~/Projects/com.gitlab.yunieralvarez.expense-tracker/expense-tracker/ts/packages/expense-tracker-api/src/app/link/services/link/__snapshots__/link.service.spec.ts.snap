// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LinkService exchangePublicToken when user is found it should attached the user to the source 1`] = `
[
  AccountEntity {
    "createdAt": 1970-01-01T00:00:00.000Z,
    "id": undefined,
    "mask": "3379",
    "name": "Wolff, Runolfsson and Schuppe",
    "plaidId": "aBWd8ptaJw_kUJf9SQiRO",
    "source": SourceEntity {
      "accessToken": "5jOaCaBoz0oDA3dcm-ffJ",
      "accounts": [],
      "createdAt": 1970-01-01T00:00:00.000Z,
      "id": undefined,
      "institution": {
        "id": "tWLMn0svIx0Xkj43n16QV",
        "name": "Daniel Group",
        "plaidId": "PQ6prHq3oUljN9fYjoAWt",
      },
      "plaidId": "Nwk0S-3w2051b00yFB7m-",
      "requestId": "2-Rl3_pLKJ3sxzbI6I_YX",
      "transactionCursor": undefined,
      "updatedAt": 1970-01-01T00:00:00.000Z,
      "user": {
        "email": "Dakota.Ruecker35@hotmail.com",
        "id": "ocf3kxo28INwBsWaXH0xH",
        "password": "4ihn1bKgs39te3GayvkFz",
      },
      Symbol(EntityRepositoryType): [Function],
    },
    "subtype": "gjr8WElkk9v4DW3_1lK4u",
    "transactions": [],
    "updatedAt": 1970-01-01T00:00:00.000Z,
    Symbol(EntityRepositoryType): [Function],
  },
  AccountEntity {
    "createdAt": 1970-01-01T00:00:00.000Z,
    "id": undefined,
    "mask": "3688",
    "name": "Dietrich and Sons",
    "plaidId": "j5_CIYzsRjl3Re2-1ICva",
    "source": SourceEntity {
      "accessToken": "5jOaCaBoz0oDA3dcm-ffJ",
      "accounts": [],
      "createdAt": 1970-01-01T00:00:00.000Z,
      "id": undefined,
      "institution": {
        "id": "tWLMn0svIx0Xkj43n16QV",
        "name": "Daniel Group",
        "plaidId": "PQ6prHq3oUljN9fYjoAWt",
      },
      "plaidId": "Nwk0S-3w2051b00yFB7m-",
      "requestId": "2-Rl3_pLKJ3sxzbI6I_YX",
      "transactionCursor": undefined,
      "updatedAt": 1970-01-01T00:00:00.000Z,
      "user": {
        "email": "Dakota.Ruecker35@hotmail.com",
        "id": "ocf3kxo28INwBsWaXH0xH",
        "password": "4ihn1bKgs39te3GayvkFz",
      },
      Symbol(EntityRepositoryType): [Function],
    },
    "subtype": "ZRG-35VCu6YOTT62iN0G1",
    "transactions": [],
    "updatedAt": 1970-01-01T00:00:00.000Z,
    Symbol(EntityRepositoryType): [Function],
  },
]
`;
